### PR TITLE
Enhancement: Extract constants to avoid mutable global state

### PR DESCRIPTION
**Pull request by localheinz**
This pull request

- [x] extracts constants to avoid mutable global state

### CODE REVIEW

**Comment by kamil-tekiela** [comment]
The indentation is off.

File: manual/spam_challenge.php [hunk 1]
```diff
@@ -27,18 +32,24 @@ function print_prefix($name, $a, $b) {
...
-	array('max',   'print_prefix'),
-	array('min',   'print_prefix'),
-	array('minus', 'print_infix', 'gen_minus'),
-	array('plus',  'print_infix', 'gen_plus'),
-);
-
+/**
+ * @return array<int, string>
+ */
+function challenges(): array {
```
Comment (localheinz) [inline manual/spam_challenge.php:38]: Running

```shell
git grep -iw challenges
```

on current `master` yields

```
archive/2012.php:991:      <li>find working solutions for your day to day challenges;</li>
archive/2013.php:378:      challenges, and we want to talk about those too. In short, this conference is a
archive/2013.php:417:      and network with peers to solve PHP challenges
archive/entries/2015-09-30-1.xml:29:     challenges. Each evening we provide a social event for people to
archive/entries/2018-06-14-1.xml:14:     <p>New season and new challenges! As conference organisers we understand perfectly that you can stay with peleton only when you focus on development. That is why phpCE, as an event aimed at a wide group of PHP programmers from Central Europe, leaves Poland for the first time. We are stronger than before thanks to the organisers of Brno PHP Conference and volunteers from the Pehapkaři group. Together we have been working for the success of this year’s edition and <strong>we are inviting you to Prague.</strong><br/></p>
include/pregen-news.inc:6835:     <p>New season and new challenges! As conference organisers we understand perfectly that you can stay with peleton only when you focus on development. That is why phpCE, as an event aimed at a wide group of PHP programmers from Central Europe, leaves Poland for the first time. We are stronger than before thanks to the organisers of Brno PHP Conference and volunteers from the Pehapkaři group. Together we have been working for the success of this year’s edition and <strong>we are inviting you to Prague.</strong><br/></p>
include/pregen-news.inc:6856:    'intro' => '<div><p>New season and new challenges! As conference organisers we understand perfectly that you can stay with peleton only when you focus on development. That is why phpCE, as an event aimed at a wide group of PHP programmers from Central Europe, leaves Poland for the first time. We are stronger than before thanks to the organisers of Brno PHP Conference and volunteers from the PehapkaÅi group. Together we have been working for the success of this yearâs edition and <strong>we are inviting you to Prague.</strong><br/></p><p style="text-align:center;font-size:120%">Our Special Guest this year will be <strong>Rasmus Lerdorf</strong>, The PHP Language Creator.</p><p class=\'fullArticleLink\'><a href=\'https://cfp.phpce.eu/\' class>&hellip; read full article</a></p></div>',
include/pregen-news.inc:20060:     challenges. Each evening we provide a social event for people to
manual/spam_challenge.php:2:// simple and stupid SPAM protection (using little challenges)
manual/spam_challenge.php:30:$challenges = array(
manual/spam_challenge.php:41:	global $challenges, $nums;
manual/spam_challenge.php:42:	$c = $challenges[rand(0, count($challenges)-1)];
manual/spam_challenge.php:55:	global $challenges, $nums;
manual/spam_challenge.php:57:	foreach ($challenges as $x) {
```

**Comment by cmb69** [comment]
Getting rid of global state is great, but why not use constants instead of variables or functions in this case?

**Comment by kamil-tekiela** [comment]
Constants are an improvement over globals, but we need some kind of encapsulation here. A class would be suitable, but maybe doing this in two-step approach is fine as we need to write tests for this too.

**Comment by localheinz** [comment]
@kamil-tekiela 

> Constants are an improvement over globals, but we need some kind of encapsulation here. A class would be suitable, but maybe doing this in two-step approach is fine as we need to write tests for this too.

That was exactly my intention!

🤓

File: manual/spam_challenge.php [hunk 2]
```diff
@@ -27,44 +27,40 @@ function print_prefix($name, $a, $b) {
...
 	array('minus', 'print_infix', 'gen_minus'),
 	array('plus',  'print_infix', 'gen_plus'),
 );
 
-
 // generate a challenge
 function gen_challenge() {
-	global $challenges, $nums;
-	$c = $challenges[rand(0, count($challenges)-1)];
+    $c = CHALLENGES[rand(0, count(CHALLENGES)-1)];
```
Comment (cmb69) [inline manual/spam_challenge.php:40]: I think we should stick with the tab indentation *for now*:
```suggestion
	$c = CHALLENGES[rand(0, count(CHALLENGES)-1)];
```

**Review by cmb69 — COMMENTED** [review]
Other than the indentation nit, this looks like a reasonable improvement per se. After all, we reduce potentially *mutable* global *state*. Putting the functionality under test can be done later.

**Review by kamil-tekiela — APPROVED** [review]

**Comment by cmb69** [comment]
Thank you!

**Comment by localheinz** [comment]
Thank you, @cmb69 and @kamil-tekiela!

**Pull request merged**